modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java

```diff
@@ -148,4 +156,23 @@ public GridNioSession session() {
     @Override public String toString() {
         return S.toString(GridTcpNioCommunicationClient.class, this, super.toString());
     }
+
+    public void sendHeartbeatsIfNeeded() {
+        if (!useHeartbeats)
+            return;
+
+        long sinceLastHeartbeat = U.currentTimeMillis() - ses.lastHeartbeat();
+
+        if (sinceLastHeartbeat > HEARTBEAT_FREQUENCY) {
+            HeartbeatMessage msg = new HeartbeatMessage();
+
+            //if (log.isDebugEnabled())
+                log.info("Heartbeat message sent [rmtAddr=" + ses.remoteAddress() + "], sinceLastHeartbeat=" + sinceLastHeartbeat
+                        + " lastSent=" + ses.lastSendTime() + " lastReceived=" + ses.lastReceiveTime() + " scheduletTime=" + ses.lastSendScheduleTime());
+
```

Comment:
This method is only called once on first connection message. Shouldn't it be called periodically during all lifetime of a connection?